### ISSUE TITLE
Rules Wishlist

### ISSUE THREAD

**Issue by federicobond** [duaraghav8/Solium#44]
This is my wishlist for additional rules. Let me know if you are against any of them getting implemented in the core. I will work on some of them but anything else is up for grabs.

 * [ ] Suggest `send` instead of `vall.value()` (WIP @federicobond)
 * [ ] Suggest avoiding `block.timestamp` as it can be manipulated by miners (only in functions) (WIP @federicobond)
 * [ ] Ensure any function that uses `msg.value` has payable modifier (WIP @federicobond)
 * [ ] Ensure fallback function has payable modifier (WIP @federicobond)
 * [ ] Ensure `send` return value is checked.
 * [ ] Suggest using `selfdestruct` instead of deprecated `suicide` (WIP @federicobond)
 * [ ] Suggest better storage packing (see Warning [here](http://solidity.readthedocs.io/en/develop/miscellaneous.html))
 * [ ] Find duplicate code (this one is hard, perhaps we can write a generic implementation and then integrate it into solium)

**Comment by maraoz** [comment]
> Ensure fallback function has payable modifier
Not always desirable. See for example: https://github.com/OpenZeppelin/zeppelin-solidity/blob/master/contracts/Rejector.sol

**Comment by federicobond** [comment]
Updated. Thanks @maraoz!

**Comment by duaraghav8** [comment]
Agreed with maroz and no issues with the rest.

My plan is to separate the linter into 2 layers - the base would only ensure compliance with the [Solidity Style Guide](https://solidity.readthedocs.io/en/develop/style-guide.html).

The 2nd layer contains all the additional rules (like you've stated above) coming from the community. But this layer is optional and therefore can be enabled / disabled by the user at any time.

Until now, I've only been working with the Base. I think we should follow the 2-layer approach.
What do you think

**Comment by federicobond** [comment]
Eslint has some notion of rulesets. See [here](http://eslint.org/docs/rules/). Perhaps we can do something similar with Solium rules, where you can choose to enable the `base` or `core` ruleset or something more complete.

I must confess I am not a big fan of the current `--sync` implementation. I believe that most users will forget to do that and lose out on newly developed rules. It's better to allow users to define a base ruleset and add/remove rules as overrides to the base ruleset(s).

**Comment by duaraghav8** [comment]
You're right, we should deprecate ```--sync``` and have a configuration file system instead.
Something similar to ESLint's.

I'll blog the final plan before I start implementing it but let me know if you have ideas of your own.

**Comment by duaraghav8** [comment]
Also Rule proposal:

- [ ] Throw **Warning** if a solidity file does not include a PRAGMA directive on top

**Comment by federicobond** [comment]
Rule:
 * [ ] Insecure instantiation of array or mapping

See: https://ethereum.karalabe.com/talks/2016-hackethon.html#10

**P.S.** Is there a secure case for this construct? Or are all its uses insecure?

**Comment by federicobond** [comment]
* [ ] Contract uses `send` or `call.value` but has no payable functions or constructor

**Comment by federicobond** [comment]
* [ ] Warn when using `tx.origin` instead of `msg.sender`

**Comment by federicobond** [comment]
* [ ] Warn about using `a =+ b` or `a =- b`. See [here](https://github.com/ether-camp/virtual-accelerator/issues/8).

As a general rule, a linter for smart contracts should be very strict with spacing around all keywords, operators, etc. to avoid these problems.

**Comment by federicobond** [comment]
* [ ] Find similar identifiers that could be confused

**Comment by Sjors** [comment]
- [ ] require explicit `public`, `private`, `external` or `internal`

See ConsenSys [best practices](https://github.com/ConsenSys/smart-contract-best-practices#explicitly-mark-visibility-in-functions-and-state-variables). Or is there any situation where this is not possible?

**Comment by Sjors** [comment]
- [ ] a function with `returns` should actually `return` for each code path, and should return the correct type
- [ ] a function without `returns` should not have any `return` statement

**Comment by duaraghav8** [comment]
All security-related rules finalized in this thread are being implemented at https://github.com/duaraghav8/solium-plugin-security

**Comment by mushketyk** [comment]
> Ensure send return value is checked

I was thinking about implementing this rule, and it seems that to avoid false positives we should keep track of all variables of type `address` and then check if method `send` is called on any of them. But if we are processing a contract that inherits from other contracts we would first have to get all variables of type `address` in all parent contracts. To do this, we need to ensure that a rule is processing parent contracts first, collect a set of addresses in parent contracts, and then check if `send` is called on any of addresses.

Do you think Solium should support something like this?
Am I overthinking this?

**Comment by duaraghav8** [comment]
@mushketyk you're not overthinking. There are several rules which are on pause because of this problem of Solium only being file-aware instead of projects-aware, ie, at any given time, a rule only has context of 1 file whereas it should have context of the entire project so any external elements can also be considered. This problem is well documented in #83 

Let's keep this rule on hold for now, because making solium project-aware is a major design change so it will take time. I'll work on it in near future.
